### ISSUE TITLE
Cloud-init drive delete when cloning vm with cloud-init drive

### ISSUE THREAD

**Issue by cwilliams001** [Telmate/terraform-provider-proxmox#901]
Environment:

Proxmox Version: 8.1.3
Terraform Provider: thegameprofi/proxmox
Provider Version: 2.10.0

Description:

I am using the @TheGamerProfi fork of the Proxmox provider due to compatibility issues with Proxmox version 8.1.3. While this fork solves the plugin crash, I have encountered a specific issue related to cloning VM templates that have a cloud-init drive attached.

Issue:

When creating a new VM by cloning a template with an attached cloud-init drive, the cloud-init drive gets deleted after the VM is created and following its move and resize operations.

Workaround:

Based on the discussion in issue #704, I applied a workaround that involves ensuring the disk type in the Terraform resource matches the disk type in the VM template. Specifically, if the VM template uses a virtio disk, the newly created VM should also use a virtio disk type to avoid boot loop issues. Additionally, I added cloudinit_cdrom_storage = "local-lvm" to the configuration.

Current Working Configuration:
```
resource "proxmox_vm_qemu" "hashi-server" {
  count       = var.server_count
  name        = "hashi-server-0${count.index + 1}"
  desc        = "hashi-server${count.index + 1}"
  vmid        = var.server_vmid_begin + count.index + 1
  target_node = var.proxmox_host
  onboot      = true
  clone    = var.template_name
  cloudinit_cdrom_storage = "local-lvm"
  full_clone  = true
  agent    = 1
  os_type  = "cloud-init"
  cores    = var.server_cores
  sockets  = 1
  cpu      = "host"
  memory   = var.server_memory
  scsihw   = "virtio-scsi-pci"
  bootdisk = "scsi0"
  disks {
    virtio {
      virtio0 {
        disk {
          size = var.server_disk_size
          storage = var.proxmox_storage
        }
      }
    }
  }
  network {
    bridge = var.proxmox_network_device
    model  = "virtio"
  }

  ipconfig0 = "ip=10.70.7.10${count.index + 1}/16,gw=10.70.0.1"
  # sshkeys   = <<EOF
  #   ${var.ssh_key}
  #   EOF
}
```

Steps to Reproduce:

Clone a VM template with an attached cloud-init drive using the specified Terraform provider and version.

Observe the deletion of the cloud-init drive post-creation and resizing of the VM

Expected Behavior:

The cloud-init drive should remain attached to the VM after cloning and resizing processes are completed.

Actual Behavior:

The cloud-init drive is removed after the VM creation process, leading to further issues such as boot loops unless the workaround is applied.

**Comment by TheGameProfi** [comment]
I can replicate the issue, and try to figure out why its happening.

**Comment by TheGameProfi** [comment]
I was able to get a closer look whats happening.

There is some kind of update of the VM after the rezize:
`update VM 104: -agent 1 -bios seabios -cores 1 -cpu host -description terraform test -hotplug network,disk,usb -ide2 none,media=cdrom -kvm 1 -memory 2048 -name `
Could be related to the changes of this commit:
[90e6dba2013de3aeb4db8395daa3fad371b9d621](https://github.com/TheGameProfi/proxmox-api-go/commit/90e6dba2013de3aeb4db8395daa3fad371b9d621)

**Comment by Tinyblargon** [comment]
@TheGameProfi could you provide the following:

- Screenshot or api data of the vm hardware
- Which storage backend is used, (LVM, File, ZFS)

**Comment by mleone87** [comment]
@TheGameProfi @Tinyblargon confirm, disk gets removed after the resize, probabily the ide2 is not parsed correctly
![Screenshot 2024-01-19 at 12 19 09](https://github.com/Telmate/terraform-provider-proxmox/assets/807457/e7915c87-ddf9-4061-9564-aab2110a2250)

**Comment by mleone87** [comment]
@Tinyblargon 
https://github.com/Telmate/proxmox-api-go/blob/cd419d1e45db3a4aa25ca0a20bbe26d9b0428aad/proxmox/config_qemu_disk_ide.go#L106

not sure what's happening there since the function is too complicate but it's basically completely ignoring the cloudinit disk got from the api response after the clone.
I also find cloudinit called as ide3 but this is not correct, it should be ide2

**Comment by Tinyblargon** [comment]
@mleone87 gonna run some tests, as I'm unsure if the bug is in Terraform > proxmox-go-api or in proxmox-go-api > proxmox.

**Comment by mleone87** [comment]
@Tinyblargon I "fixed" it modifying the api, as far as I can tell the proxmox code is doing well in that part

**Comment by Tinyblargon** [comment]
@mleone87 do you mean the `proxmox-go-api`?

**Comment by mleone87** [comment]
@Tinyblargon yes Sorry, I'll submit a PR later to show

**Comment by Tinyblargon** [comment]
@mleone87 
How it was intended to work is 

1. Clone vm
2. Get vm config
3. Delete and add disks
4. Start create cloudinit config
5. Start vm

I think I've figured out what is happening. When the removal of the `cloud-init` disk becomes a pending change, you can't add another disk as technically the old `cloud-init` disk is still there.
To fix this, we would first have to delete it, then add disks.

So step 3 should become delete then add.

When my template doesn't have a `cloud-init` disk or it's there as `ide3`, it is created/remains as `ide3`.

**Comment by cosminmocan** [comment]
Just stumbled upon this issue, and boy is it annoying, first i had an issue where my disk kept on getting unmounted when trying to mount it using virtio instead of scsi0, now this is the next challenge :)) .
What I have noticed is that the example directory also has a file for the cloud-init example that is 3 years old, with the old disk definition.
As for this issue, I am currently using the compiled master branch plugin, is this bug already solved on another tag/branch or should we wait a bit.

Huge thank you for everyone involved in the creation and maintenance of this awesome provider!

**Comment by Tinyblargon** [comment]
@mleone87 okay so setting `ciuser` fixed it for me, this is definitely an issue with the mapping between Terraform and proxmox-go-api. https://github.com/Telmate/terraform-provider-proxmox/blob/47df793ca73c899fd599975e023de081ca86fac4/proxmox/resource_vm_qemu.go#L2336-L2344

On another note https://github.com/Telmate/proxmox-api-go/issues/299 fixes an issue you run into when you set `ciuser`. We try to add an new `cloud-init` disk while the old one technically still exists, since there may only be one `cloud-init` disk updating fails.

**Comment by mleone87** [comment]
@Tinyblargon in my template I have the `ciuser` but I usually do not set the `cloudinit_cdrom_storage` that make that block effective, so it should be set as a default and only explicitly modified if needed?

Also on line 2339 you set cdrom id to be ide3 but as per proxmox api docs configuration it should be ide2, at least for the first cdrom set

**Comment by Tinyblargon** [comment]
@mleone87 Well, since `cloudinit_cdrom_storage` is dependent on the names of the available storages, it's best to leave this empty by default. The easiest solution will be to add to the docs that `cloudinit_cdrom_storage` should be specified along with a `cloud-init` setting to add the `cloud-init` disk.

**Comment by mleone87** [comment]
@Tinyblargon there is still something that does not compute well around cloudinit drive, at least for me.

- first of all I'm a very big fan of KISS 😄 
- second, we are ignoring what we got from the API and ask user to explicitly set a storage when is not mandatory, as I remembere the cloudinit_cdrom_storage was introduced to diff from _default same storage as disk0_
- third, if we set cloud-int disk to be ide3 and iso disk to be ide2 then we end up with a configuration that we did not explicitly set anywhere, and is different from clone source
![Screenshot 2024-01-23 at 16 47 04](https://github.com/Telmate/terraform-provider-proxmox/assets/807457/4d8f7c29-f5cc-4bdf-9089-7bcbd5e644e7)

**Comment by electropolis** [comment]
I see I'm struggling the same problem but mention that in different issue https://github.com/Telmate/terraform-provider-proxmox/issues/922

I have templated with ide2 as cloudinit but during the setup with terraform it disappeared. Everything was fine when 2.9.3 provider used with Proxmox v7.1. Suddenly everything crashed when Proxmox v8.1.3 came with the latest provider. So what is the solution here guys because I'm confused. Maybe a solution? Because all tutorials for creating [template](https://pve.proxmox.com/wiki/Cloud-Init_Support)  are not capable to work with terraform provider.

**Comment by devZer0** [comment]
this worked for me https://github.com/Telmate/terraform-provider-proxmox/issues/704#issuecomment-1936723112

**Comment by electropolis** [comment]
> this worked for me [#704 (comment)](https://github.com/Telmate/terraform-provider-proxmox/issues/704#issuecomment-1936723112)

I will check that but there is scenario about not attaching cloudinit to template at all and I think and I already tried that and didn't worked

**Comment by devZer0** [comment]
i have not attached cloudinit to template at all. it just works after adding this param

**Comment by electropolis** [comment]
> i have not attached cloudinit to template at all. it just works after adding this param

You mean when cloning the template and setting the cloudinit drive ?

**Comment by devZer0** [comment]
i don't need to set a cloudinit drive, i only have

	os_type  = "cloud-init"
	cloudinit_cdrom_storage = "local-zfs"
	
a cloudinit ide drive magically appears without doing any further.

**Comment by electropolis** [comment]
> i don't need to set a cloudinit drive, i only have
> 
> ```
> os_type  = "cloud-init"
> cloudinit_cdrom_storage = "local-zfs"
> ```
> 
> a cloudinit ide drive magically appears without doing any further.

Yeah but do you set the cloudinit also in template? I'm talking about Template preparing procedure 
https://pve.proxmox.com/wiki/Cloud-Init_Support#_preparing_cloud_init_templates
its described here and there is chapter `Add Cloud-Init CD-ROM drive`

_The next step is to configure a CD-ROM drive, which will be used to pass the Cloud-Init data to the VM._
`qm set 9000 --ide2 local-lvm:cloudinit`

**Comment by devZer0** [comment]
i have no cloudinit drive in template. it simply works without, as cloudinit drive seems to be added by terraform

**Comment by electropolis** [comment]
>[...] nor do i add cdrom/cloudinit drive in terraform configuration. 

How's so ? You add it here

> i don't need to set a cloudinit drive, i only have
> 
> ```
> os_type  = "cloud-init"
> cloudinit_cdrom_storage = "local-zfs"
> ```

That's cloudinit added in terraform.

**Comment by electropolis** [comment]
@devZer0 

I created template without cloudinit image

![image](https://github.com/Telmate/terraform-provider-proxmox/assets/16556636/22d8baa3-0a7d-4977-b616-b8603d1aa036)


I hashed the code in ansible responsible for creating the image
```
    # - name: Settings for Cloudinit
    #   tags: cloudinit
    #   block:

    #     - name: Set fact for cloudinit disk to check if exists
    #       ansible.builtin.set_fact:
    #         cloudinit_image: "{{ cloudinit_image | default([]) + [item] }}"
    #       loop: "{{ qm_config.results }}"
    #       when: not item.stdout is search('vm-{{ item.item.template_id}}-cloudinit')

    #     - name: CloudInit results
    #       ansible.builtin.debug:
    #         var: cloudinit_image
    #         verbosity: 2

    #     - name: Add cloud-init image as CDROM
    #       ansible.builtin.command: "qm set {{ item.item.template_id }} --ide3 local-lvm:cloudinit"
    #       loop: "{{ cloudinit_image }}"
    #       when: cloudinit_image is defined
```

When running terraform it sticks on `Still creating`
And I Proxmox I observe this 

![image](https://github.com/Telmate/terraform-provider-proxmox/assets/16556636/cd287174-4cbd-4ddc-8ae6-f98a8e028f41)

no cloutdinit drive only CD-ROM on ide2.

**Comment by electropolis** [comment]
> In my experiments when I add this to my terraform code:
> 
> ```
>   ciuser   = "ciuser_name"
>   cipassword = "<some_password>"
>   cloudinit_cdrom_storage = "local-lvm"
> ```
> 
> AND
> 
> I use a VM template without a cloudinit drive for cloning.
> 
> Then I get a VM that has empty CD drive on IDE2 and cloud-init drive on IDE3. Of course this setup fails to boot as cloud-init drives are expected to be on IDE2. When I manually delete both IDE drives and create cloud-init drive on IDE2 it works.
> 
> No idea if there is a way to force this provider to create cloud-init drive on IDE2.

So definitely an issue. I'm using `cicustom`  because I prefer to have cloudinit on VM with specific settings instead using limited `ci<parameter>` that doesn't have many other options to set it up. And that setup also doesn't work..  
I think there is huge mess of information and now there are at least 3-4 issues that are describing that error. There are even some person who say that it works. I don't know how although they also have IDE2 with empty CD-ROM

**Comment by vasekhodina** [comment]
> So definitely an issue. I'm using `cicustom` because I prefer to have cloudinit on VM with specific settings instead using limited `ci<parameter>` that doesn't have many other options to set it up. And that setup also doesn't work.. I think there is huge mess of information and now there are at least 3-4 issues that are describing that error. There are even some person who say that it works. I don't know how although they also have IDE2 with empty CD-ROM

I agree it's a mess now. I think those people that say it works didn't test if the machine boots. Just that the cloudinit drive is there was enough for them.

**Comment by cwilliams001** [comment]
@vasekhodina the machines do in fact boot. If you follow the steps for the work around correctly you should be able to have a working template.

Create a template with cloud init.

After the template is made delete the cloud init drive from the template in the hardware settings of the web UI.

Regenerate the image for your template.

Clone your template


Start VM

If this doesn't work there is a mismatch of something in your terraform code. Something probably with the disk type expected and which one is actually booting.


If you jump in the discord there are others who have working configs and are more than happy to help, I agree with you that this is an issue. I disagree with your implications that people are just saying it works so they can write about it here.

**Comment by cwilliams001** [comment]
@vasekhodina  @electropolis [https://discord.gg/makTPGF3Zw](discord link)

@electropolis I mispoke I apologize I regenerate the cloud init image for nothing but good measure before I detach the cloud init drive from the template. You are also correct in this is an issue with pve version 8 and above because of changes they have made to the api. The community though as you'll see in the discord has really revived this project int he last few weeks and are working to fix bugs like these. It is by no means a perfect solution and I am a sample size of one but there are others who have been able to use the provider successfully. 

My current working terraform file

```
resource "proxmox_vm_qemu" "kube-master" {
  count                   = var.server_count
  name                    = "kube-master-0${count.index + 1}"
  desc                    = "kube-master${count.index + 1}"
  vmid                    = var.server_vmid_begin + count.index + 1
  target_node             = var.proxmox_host
  onboot                  = true
  clone                   = var.template_name
  cloudinit_cdrom_storage = var.proxmox_storage
  full_clone              = true
  agent                   = 1
  os_type                 = "cloud-init"
  cores                   = var.server_cores
  sockets                 = 1
  cpu                     = "host"
  memory                  = var.server_memory
  scsihw                  = "virtio-scsi-pci"
  bootdisk                = "scsi0"
  disks {
    scsi {
      scsi0 {
        disk {
          size       = var.server_disk_size
          storage    = var.proxmox_storage
          emulatessd = true
        }
      }
    }
  }
  network {
    bridge = var.proxmox_network_device
    model  = "virtio"
  }

  ipconfig0 = "ip=10.70.7.10${count.index + 1}/16,gw=10.70.0.1"
  sshkeys   = <<EOF
    ${var.pub_ssh_key}
    EOF

  lifecycle {
    ignore_changes = [
      network,
      ciuser,
      qemu_os
    ]
  }


}
resource "proxmox_vm_qemu" "kube-worker" {
  count                   = var.client_count
  name                    = "kube-worker-0${count.index + 1}"
  desc                    = "kube-worker${count.index + 1}"
  vmid                    = var.client_vmid_begin + count.index + 1
  target_node             = var.proxmox_host
  onboot                  = true
  full_clone              = true
  clone                   = var.template_name
  cloudinit_cdrom_storage = var.proxmox_storage
  agent                   = 1
  os_type                 = "cloud-init"
  cores                   = var.client_cores
  sockets                 = 1
  cpu                     = "host"
  memory                  = var.client_memory
  scsihw                  = "virtio-scsi-pci"
  bootdisk                = "scsi0"
  disks {
    scsi {
      scsi0 {
        disk {
          size       = var.client_disk_size
          storage    = var.proxmox_storage
          emulatessd = true
        }
      }
    }
  }

  network {
    bridge = var.proxmox_network_device
    model  = "virtio"
  }

  ipconfig0 = "ip=10.70.8.10${count.index + 1}/16,gw=10.70.0.1"
  sshkeys   = <<EOF
    ${var.pub_ssh_key}
    EOF

  lifecycle {
    ignore_changes = [
      network,
      ciuser,
      qemu_os
    ]
  }
}


resource "local_file" "ansible_inventory" {
  content = templatefile("inventory.tmpl",
    {
      servers = tomap({
        for idx in range(var.server_count) :
        "kube-master-${format("%02d", idx + 1)}" => split("/", split("=", split(",", proxmox_vm_qemu.kube-master[idx].ipconfig0)[0])[1])[0]
      })
      clients = tomap({
        for idx in range(var.client_count) :
        "kube-worker-${format("%02d", idx + 1)}" => split("/", split("=", split(",", proxmox_vm_qemu.kube-worker[idx].ipconfig0)[0])[1])[0]
      })
    }
  )
  filename = "../inventory"
}
```
![vm](https://github.com/Telmate/terraform-provider-proxmox/assets/82992893/e1c91088-3437-4090-af3f-65202657e225)
![template](https://github.com/Telmate/terraform-provider-proxmox/assets/82992893/0485d30e-3772-4328-8213-d34c7fa8b58d)

**Comment by vasekhodina** [comment]
Ok, so after I redefined by boot order the machine started.
Thanks to @cwilliams001 for encouraging me to check my config again.

So here's a summary of what to do, in case you're from the future and you ran into the same issue as I did.
 
>1. Create a template with cloud init.
> 
>2. After the template is made delete the cloud init drive from the template in the hardware settings of the web UI.
> 
>3. Regenerate the image for your template.
> 
>4. Clone your template, etc. (We want to do this via Terraform.

Basically you want to have a template with cloudinit image ready but *without* the cloud-init drive. If the cloud init drive was present it would just get deleted not created. You'd end up with empty IDE2 drive.

Next you want to make sure you have the following values in your Terraform code as arguments for "proxmox_vm_qemu" resource.
```
# These 3 vars are there so that cloud-init drive gets created
  ciuser   = "ciuser_name"
  cipassword = "<some_password>"
  cloudinit_cdrom_storage = "local-lvm"
# The following is for making sure that when the VM get's created it knows how to boot
  boot                    = "order=scsi0;ide3"
```
What happens is that when the drive is missing after cloning it will get added by Terraform as IDE3 CD drive alongside with empty IDE2 drive. That's why we need to define the bootorder to contain the IDE3 drive. Otherwise IDE2 might get used by default and the boot process might fail.

Of course if you have your boot disk connected via any other way that scsi, you need to change the first entry in boot order. My boot disk is connected via scsi.

**Comment by electropolis** [comment]
ok @vasekhodina you point some important and crucial things than none of commentators described
>Basically you want to have a template with cloudinit image ready but without the cloud-init drive. If the cloud init drive was present it would just get deleted not created. You'd end up with empty IDE2 drive.

So it means you can have any cloudinit drive on ide2 but it doesn't mean you can't have it at all in the template unless it's configured on IDE2. Having cloudinit on IDE2 should work. I have tested both cases. WITH and WITHOUT none of them worked. But you mention different case further. 

```
# These 3 vars are there so that cloud-init drive gets created
  ciuser                  = "ciuser_name"
  cipassword              = "<some_password>"
  cloudinit_cdrom_storage = "local-lvm"
```
Nah, I don't agree. Although `cloudinit_cdrom_storage = "local-lvm"` is crucial and some people don't have and some do. I asked that already here in some Issues, and author said that `cloudinit_cdrom_storage` should be always set. But those options are for static cloudinit settings when you want them to be set from GUI here
<img width="603" alt="image" src="https://github.com/Telmate/terraform-provider-proxmox/assets/16556636/51639f3d-027e-4d87-818d-6097af6b64e4">

Which is described here https://pve.proxmox.com/wiki/Cloud-Init_Support#_deploying_cloud_init_templates
but mine config doesn't rely on those settings as they are very limited. Instead I rely on this https://pve.proxmox.com/wiki/Cloud-Init_Support#_custom_cloud_init_configuration and that is why I'm using `cicustom` 
https://github.com/sonic-networks/terraform/blob/master/proxmox/sonic/main.tf#L32

It's more flexible and worked - in pve7. 
But as I said you mention interesting thing:  `boot                    = "order=scsi0;ide3"`
But this is only for booting to have cloudinit booted yes? But what about cloudinit that should use snippet ? :/

**Comment by vasekhodina** [comment]
@electropolis Looks like the way you use proxmox is a lot different from mine. This around my 7th day of using proxmox and terraform so I'm not an expert and I'm glad I made the stuff work.
So I can't help you, I'm sorry. Hope you have luck on the Discord channel.

**Comment by electropolis** [comment]
> @electropolis Looks like the way you use proxmox is a lot different from mine. This around my 7th day of using proxmox and terraform so I'm not an expert and I'm glad I made the stuff work. So I can't help you, I'm sorry. Hope you have luck on the Discord channel.

How I can use proxmox different way? The goal is the same. But I want to have cicustom with real cloudinit like it is done in the cloud providers. That's an option that is provided by proxmox and can be used.

**Comment by hestiahacker** [comment]
I've confirmed that this still affects 3.0.1-rc1.

I've also confirmed that what @vasekhodina said about [setting the boot order](https://github.com/Telmate/terraform-provider-proxmox/issues/901#issuecomment-1939589804) does work around this issue. I don't entirely understand why, but it causes the cloud-init drive to be attached to ide3 (which is available) instead of ide0 (which is occupied).

What @electropolis said about `cloudinit_cdrom_storage` [being required](https://github.com/Telmate/terraform-provider-proxmox/issues/901#issuecomment-1939613477) is correct. This was mentioned in #935 and I submitted a merge request to get the documentation updated: https://github.com/Telmate/terraform-provider-proxmox/pull/939

To @electropolis's question about whether cloud-init works in this provider without cicustom, I can attest that it does. Below are the relevant settings that I use to set a static IP address and specify nameservers on my machines. If you're terraform looks similar and you're still having trouble, please post a minimal terraform which has the issue and I'll use that to try to reproduce the issue on my side.

```
  # Cloud Init Settings
  cloudinit_cdrom_storage = var.storage_backend
  # Reference: https://pve.proxmox.com/wiki/Cloud-Init_Support
  ipconfig0 = "ip=${var.ip_address}/${var.cidr},gw=${var.gateway}"
  nameserver = var.nameservers
  sshkeys = var.sshkeys
  boot = "order=virtio0;ide3"
```

While I'm currently testing with 3.0.1-rc1 of the provider, I had cloud init working with 2.9.11 as well (without using cicustom).

**Comment by electropolis** [comment]
@hestiahacker but it doesn't work with cicustom which means there is a bug still.

**Comment by hestiahacker** [comment]
Yeah, you're right. I thought you were trying to use cicustom to work around this issue, but now that I've re-read the thread more closely I see that's not the case. Also, even if you were just trying to work around something the fact that cicustom doesn't work is still a problem, so I'm not sure what I was thinking. :facepalm: 

I went through the other 5 open tickets that mention cicustom and it seemed to only be mentioned in passing or in examples. I didn't see a ticket just saying "cicustom doesn't work" which seems to be the case here. If you'd be willing to make a new ticket about this and tag me in a comment right away so I don't miss it, I'll put this in my queue of things to look into.

I tried using it something like a year ago and couldn't get it to work, but I thought it was just because I didn't know what I was doing. I'd probably also use cicustom if I better understood what it can do, and the more people using it, the more likely we are to find and fix these issues quickly.

**Comment by electropolis** [comment]
> Yeah, you're right. I thought you were trying to use cicustom to work around this issue, but now that I've re-read the thread more closely I see that's not the case. Also, even if you were just trying to work around something the fact that cicustom doesn't work is still a problem, so I'm not sure what I was thinking. 🤦
> 
> I went through the other 5 open tickets that mention cicustom and it seemed to only be mentioned in passing or in examples. I didn't see a ticket just saying "cicustom doesn't work" which seems to be the case here. If you'd be willing to make a new ticket about this and tag me in a comment right away so I don't miss it, I'll put this in my queue of things to look into.
> 
> I tried using it something like a year ago and couldn't get it to work, but I thought it was just because I didn't know what I was doing. I'd probably also use cicustom if I better understood what it can do, and the more people using it, the more likely we are to find and fix these issues quickly.

I could create a new issue but its still copies everything again with description. I could do that but today in the evening.

**Comment by hestiahacker** [comment]
You can just make the new issue short and sweet and then link to this ticket. All I need is a minimal terraform file that demonstrates the issue and the expected/actual behavior.

I just posted a [minimal example](https://github.com/Telmate/terraform-provider-proxmox/issues/919#issuecomment-1952835555) that would make an idea starting place for me. Just change the Cloud Init Settings and I should be able to test it quickly and easily. Just to set expectations, it'll probably be next Monday until I get to it. Mondays are my "work on the proxmox provider" days.

**Comment by adamgass** [comment]
@hestiahacker I have been having issues with getting cicustom to work similar to what others have mentioned and a simple workaround I have found is to specify a ciuser or nameserver within the TF resource block.  Having something like `ciuser="default"` lets the cloud-init drive mount to the vm and cloud-init is off and running.  I think it just needs some kind of input parameter for the drive to function properly.  It's almost as if when a vm is cloned there is some type of logic written during the mount process that checks for input and if nothing is there then it discards the drive.

**Comment by electropolis** [comment]
> @hestiahacker I have been having issues with getting cicustom to work similar to what others have mentioned and a simple workaround I have found is to specify a ciuser or nameserver within the TF resource block. Having something like `ciuser="default"` lets the cloud-init drive mount to the vm and cloud-init is off and running. I think it just needs some kind of input parameter for the drive to function properly. It's almost as if when a vm is cloned there is some type of logic written during the mount process that checks for input and if nothing is there then it discards the drive.

If you wanna use `cicustom` above your comment Tinyblargon provided the link. Just follow it. There is nothing to check, try and so on. The solutions above describes how to compile new provider with changed code that omits the cdrom and it doesn't disappear.

**Comment by ggrocco** [comment]
I'm having a similar issue, it's deleting all the cloud-init info
```
Jun 26 11:39:02 pve pvedaemon[518008]: <root@pam!terraform> update VM 221: -agent 1 -balloon 0 -bios seabios -boot order=scsi0 -cicustom  -ciupgrade 0 -cores 6 -cpu host -delete cpuunits,ciuser,cipassword,searchdomain,nameserver,sshkeys,shares -description Node 01 -hotplug network,disk,usb -ipconfig0 ip=192.168.54.21/26,gw=192.168.54.1 -kvm 1 -memory 16384 -name k3s-n01 -net0 virtio,bridge=vmbr0 -numa 0 -onboot 1 -protection 0 -scsi0 local-lvm:vm-221-disk-0,discard=on,replicate=0,ssd=1 -scsihw virtio-scsi-pci -sockets 1 -tablet 1
```

<img width="538" alt="Image" src="https://github.com/user-attachments/assets/a5c4b420-aa69-4060-ae77-ee3ceea3a510" />
<img width="636" alt="Image" src="https://github.com/user-attachments/assets/ebfd6c69-a025-41a2-9ba0-a34322c69ddf" />

In the past, I was using lifecycle.ignore_changes, and it was working, now, with the upgrade, it doesn't make any difference, and everything is removed.

```
lifecycle {
    ignore_changes = [
      ciuser,
      cipassword,
      sshkeys,
      network
    ]
  }
```

Any tip how to fix this?

**Comment by ruddem** [comment]
Ensure that the Cloud-Init disk is explicitly defined in your Terraform configuration. Add the following block to your main.tf, alongside the definition of the primary (OS) disk:

```
disk {
  slot     = "ide2"
  type     = "cloudinit"
  storage  = var.storage
}
```

This ensures that the Cloud-Init drive is properly attached during the cloning process and prevents it from being removed.

Tested with:

- Telmate Proxmox Provider: 3.0.2-rc07
- Proxmox VE: 9.1.7
